File: salesforce-okta/src/main/java/com/jinloes/salesforce/web/SalesforceController.java
package com.jinloes.salesforce.web;

import com.jinloes.salesforce.client.SalesforceClient;
import com.jinloes.salesforce.model.SalesforceQueryResult;
import com.jinloes.salesforce.service.SalesforceTokenService;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/salesforce")
@RequiredArgsConstructor
public class SalesforceController {

  private final SalesforceTokenService tokenService;
  private final SalesforceClient salesforceClient;

  @GetMapping("/query")
  public SalesforceQueryResult query(
      @RequestHeader("Authorization") String authHeader, @RequestParam String soql) {
    var token = tokenService.exchange(extractBearerToken(authHeader));
    return salesforceClient.query(token, soql);
  }

  @GetMapping("/sobjects/{type}/{id}")
  public Map<?, ?> getRecord(
      @RequestHeader("Authorization") String authHeader,
      @PathVariable String type,
      @PathVariable String id) {
    var token = tokenService.exchange(extractBearerToken(authHeader));
    return salesforceClient.getRecord(token, type, id);
  }

  @PostMapping("/sobjects/{type}")
  public ResponseEntity<Map<String, String>> createRecord(
      @RequestHeader("Authorization") String authHeader,
      @PathVariable String type,
      @RequestBody Map<String, Object> fields) {
    var token = tokenService.exchange(extractBearerToken(authHeader));
    String id = salesforceClient.createRecord(token, type, fields);
    return ResponseEntity.ok(Map.of("id", id));
  }

  @PatchMapping("/sobjects/{type}/{id}")
  public ResponseEntity<Void> updateRecord(
      @RequestHeader("Authorization") String authHeader,
      @PathVariable String type,
      @PathVariable String id,
      @RequestBody Map<String, Object> fields) {
    var token = tokenService.exchange(extractBearerToken(authHeader));
    salesforceClient.updateRecord(token, type, id, fields);
    return ResponseEntity.noContent().build();
  }

  @DeleteMapping("/sobjects/{type}/{id}")
  public ResponseEntity<Void> deleteRecord(
      @RequestHeader("Authorization") String authHeader,
      @PathVariable String type,
      @PathVariable String id) {
    var token = tokenService.exchange(extractBearerToken(authHeader));
    salesforceClient.deleteRecord(token, type, id);
    return ResponseEntity.noContent().build();
  }

  private String extractBearerToken(String authHeader) {
    if (authHeader != null && authHeader.startsWith("Bearer ")) {
      return authHeader.substring(7);
    }
    throw new IllegalArgumentException("Missing or invalid Authorization header");
  }
}

File: salesforce-okta-csharp/Controllers/SalesforceController.cs
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;
using salesforce_okta_csharp.Services;

namespace salesforce_okta_csharp.Controllers;

[ApiController]
[Route("salesforce")]
public class SalesforceController(SalesforceTokenService tokenService, SalesforceClient salesforceClient)
    : ControllerBase
{
    [HttpGet("query")]
    public async Task<IActionResult> Query(
        [FromHeader(Name = "Authorization")] string authHeader,
        [FromQuery] string soql
    )
    {
        var token = await tokenService.ExchangeAsync(ExtractBearerToken(authHeader));
        var result = await salesforceClient.QueryAsync(token, soql);
        return Ok(result);
    }

    [HttpGet("sobjects/{type}/{id}")]
    public async Task<IActionResult> GetRecord(
        [FromHeader(Name = "Authorization")] string authHeader,
        string type,
        string id
    )
    {
        var token = await tokenService.ExchangeAsync(ExtractBearerToken(authHeader));
        var result = await salesforceClient.GetRecordAsync(token, type, id);
        return Ok(result);
    }

    [HttpPost("sobjects/{type}")]
    public async Task<IActionResult> CreateRecord(
        [FromHeader(Name = "Authorization")] string authHeader,
        string type,
        [FromBody] JsonElement fields
    )
    {
        var token = await tokenService.ExchangeAsync(ExtractBearerToken(authHeader));
        var id = await salesforceClient.CreateRecordAsync(token, type, fields);
        return Ok(new { id });
    }

    [HttpPatch("sobjects/{type}/{id}")]
    public async Task<IActionResult> UpdateRecord(
        [FromHeader(Name = "Authorization")] string authHeader,
        string type,
        string id,
        [FromBody] JsonElement fields
    )
    {
        var token = await tokenService.ExchangeAsync(ExtractBearerToken(authHeader));
        await salesforceClient.UpdateRecordAsync(token, type, id, fields);
        return NoContent();
    }

    [HttpDelete("sobjects/{type}/{id}")]
    public async Task<IActionResult> DeleteRecord(
        [FromHeader(Name = "Authorization")] string authHeader,
        string type,
        string id
    )
    {
        var token = await tokenService.ExchangeAsync(ExtractBearerToken(authHeader));
        await salesforceClient.DeleteRecordAsync(token, type, id);
        return NoContent();
    }

    private static string ExtractBearerToken(string authHeader)
    {
        if (authHeader?.StartsWith("Bearer ") == true)
            return authHeader["Bearer ".Length..];
        throw new ArgumentException("Missing or invalid Authorization header");
    }
}
